File: backend/utils/auditLogger.js
const db = require("../config/db");

/**
 * Log a security, administrative, or recruitment event in the audit trail.
 * Multi-tenant safe, append-only, and scrubs sensitive metadata properties.
 */
const logAuditEvent = async ({
    req,
    organizationId,
    actorId,
    actorName,
    actorEmail,
    eventCategory,
    action,
    resourceType,
    resourceId = null,
    result = "SUCCESS",
    metadata = {}
}) => {
    try {
        let finalOrgId = organizationId;
        let finalActorId = actorId;
        let finalActorName = actorName;
        let finalActorEmail = actorEmail;
        let finalIp = null;
        let finalUa = null;

        // If Express request is provided, extract client metadata and auth context
        if (req) {
            finalIp = (req.headers ? req.headers["x-forwarded-for"] : null) || req.socket?.remoteAddress || null;
            // Handle proxy arrays
            if (finalIp && finalIp.includes(",")) {
                finalIp = finalIp.split(",")[0].trim();
            }
            finalUa = req.headers ? req.headers["user-agent"] : null;

            if (req.user) {
                finalOrgId = finalOrgId || req.user.organization_id;
                finalActorId = finalActorId || req.user.id;
                finalActorEmail = finalActorEmail || req.user.email;
            }
        }

        // Clean metadata to protect secrets from being logged in the audit trail
        const cleanMetadata = { ...metadata };
        const sensitiveKeys = [
            "password",
            "password_hash",
            "token",
            "token_hash",
            "currentpassword",
            "newpassword",
            "confirmpassword",
            "secret",
            "apikey",
            "authorization",
            "cookie"
        ];

        for (const key of Object.keys(cleanMetadata)) {
            const lowerKey = key.toLowerCase();
            if (sensitiveKeys.some(sk => lowerKey.includes(sk))) {
                cleanMetadata[key] = "[REDACTED]";
            }
        }

        // Final fallbacks for system background jobs
        finalOrgId = finalOrgId || 1; // Default to demo org
        finalActorId = finalActorId || null; // Nullable if no user context
        finalActorEmail = finalActorEmail || "system@ats.com";
        if (!finalActorName) {
            finalActorName = finalActorEmail.split("@")[0] || "System";
        }

        const sql = `
            INSERT INTO audit_logs 
            (organization_id, actor_id, actor_name, actor_email, event_category, action, resource_type, resource_id, result, ip_address, user_agent, metadata)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        `;

        const params = [
            finalOrgId,
            finalActorId,
            finalActorName,
            finalActorEmail,
            eventCategory,
            action,
            resourceType,
            resourceId,
            result,
            finalIp,
            finalUa ? finalUa.substring(0, 255) : null,
            JSON.stringify(cleanMetadata)
        ];

        db.query(sql, params, (err) => {
            if (err) {
                console.error("CRITICAL: Failed to write audit event log:", err);
            }
        });
    } catch (err) {
        console.error("CRITICAL: Error in auditLogger utility:", err);
    }
};

module.exports = {
    logAuditEvent
};

File: backend-sb/src/main/java/com/ats/service/AuditLogger.java
package com.ats.service;

import com.ats.entity.AuditLog;
import com.ats.repository.AuditLogRepository;
import com.ats.repository.OrganizationRepository;
import com.ats.repository.UserRepository;
import com.ats.security.CustomUserDetails;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.Map;

@Service
@Transactional
public class AuditLogger {

    @Autowired
    private AuditLogRepository auditLogRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private OrganizationRepository organizationRepository;

    private static final ObjectMapper objectMapper = new ObjectMapper();

    public void logEvent(HttpServletRequest request, Integer organizationId, Integer actorId,
                         String actorName, String actorEmail, String eventCategory, String action,
                         String resourceType, Integer resourceId, String result, Map<String, Object> metadata) {
        try {
            Integer finalOrgId = organizationId;
            Integer finalActorId = actorId;
            String finalActorName = actorName;
            String finalActorEmail = actorEmail;
            String finalIp = null;
            String finalUa = null;

            if (request != null) {
                finalIp = request.getHeader("X-Forwarded-For");
                if (finalIp != null && finalIp.contains(",")) {
                    finalIp = finalIp.split(",")[0].trim();
                }
                if (finalIp == null) {
                    finalIp = request.getRemoteAddr();
                }
                finalUa = request.getHeader("User-Agent");
            }

            if (SecurityContextHolder.getContext().getAuthentication() != null) {
                Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
                if (principal instanceof CustomUserDetails) {
                    CustomUserDetails userDetails = (CustomUserDetails) principal;
                    if (finalOrgId == null) finalOrgId = userDetails.getOrganizationId();
                    if (finalActorId == null) finalActorId = userDetails.getId();
                    if (finalActorEmail == null) finalActorEmail = userDetails.getEmail();
                    if (finalActorName == null) finalActorName = userDetails.getUsername().split("@")[0];
                }
            }

            Map<String, Object> cleanMetadata = new HashMap<>();
            if (metadata != null) {
                for (Map.Entry<String, Object> entry : metadata.entrySet()) {
                    cleanMetadata.put(entry.getKey(), entry.getValue());
                }
            }
            String[] sensitiveKeys = {
                    "password", "password_hash", "token", "token_hash",
                    "currentpassword", "newpassword", "confirmpassword",
                    "secret", "apikey", "authorization", "cookie"
            };

            for (String key : cleanMetadata.keySet()) {
                String lowerKey = key.toLowerCase();
                for (String sk : sensitiveKeys) {
                    if (lowerKey.contains(sk)) {
                        cleanMetadata.put(key, "[REDACTED]");
                    }
                }
            }

            if (finalOrgId == null) finalOrgId = 1;
            if (finalActorEmail == null) finalActorEmail = "system@ats.com";
            if (finalActorName == null) {
                finalActorName = finalActorEmail.split("@")[0];
            }

            String metadataJson = objectMapper.writeValueAsString(cleanMetadata);

            AuditLog auditLog = new AuditLog();
            auditLog.setOrganization(organizationRepository.getReferenceById(finalOrgId));
            if (finalActorId != null) {
                auditLog.setActor(userRepository.getReferenceById(finalActorId));
            }
            auditLog.setActorName(finalActorName);
            auditLog.setActorEmail(finalActorEmail);
            auditLog.setEventCategory(eventCategory);
            auditLog.setAction(action);
            auditLog.setResourceType(resourceType);
            auditLog.setResourceId(resourceId);
            auditLog.setResult(result);
            auditLog.setIpAddress(finalIp);
            auditLog.setUserAgent(finalUa != null ? finalUa.substring(0, Math.min(finalUa.length(), 255)) : null);
            auditLog.setMetadata(metadataJson);

            auditLogRepository.save(auditLog);
        } catch (Exception e) {
            System.err.println("CRITICAL: Failed to write audit event log: " + e.getMessage());
            e.printStackTrace();
        }
    }
}
